### ISSUE TITLE
[rl] Fixed length train batch

### ISSUE THREAD

**Issue by joecummings** [pytorch/torchtitan#3110]
### Context

We use varlen attention for training/inference in our RL experiment. In training, we naively take all trajectories and put them together to pass to the model. This is efficient in terms of tflops since we don't waste any compute on padding tokens that would just be masked out; however, it results in variable length sequences, which triggers recompiles and is therefore NOT as efficient under a compiled regime. 

The high-level solution? We need fixed length sequences being passed through our model forward. 

### Suggestion

In order to maintain the efficiency of varlen attention, we want to pack the sequences to a maximum sequence length. 

```python
episodes = self._build_episodes(trajectories)
# Iterates over num_bins=dp_size
per_rank_episodes = self._bin_pack_episodes(episodes, self._buffer_size=seq_len)
batches = [self._collate_episodes(eps) for eps in per_rank_episodes]
```

Things to note with this approach:
* Likely there will be one final wasted sequence at the end of each pack, need to mask that
* Bins will have unequal episode counts: actual loss needs to manually scale the loss by sum of global tokens, TrainBatch should track the num episodes and num response tokens
* There will be "spillover" episodes that need to be popped and appended to the next training batch

Longer term considerations:
* Even in the sync case, we could consider something like a ReplayBuffer Actor that could sit co-located with the trainer and would handle all the packing and collating - easier conceptually and will be needed eventually for async
